---
title: "Analysis of the BUSCO reports for ferrets"
author: "Lorena_Derezanin"
output: html_document
---

## Black-footed ferret(BFF) and domestic ferret(DF) HiC assembly BUSCO-report analysis

#### BUSCO version: 3.0.2
#### lineage dataset: mammalia_odb9 (created on: 2016-02-13), number of species: 50, number of BUSCOs: 4104

``` {r renv_init_load_pkgs}

# initialize renv for a private pkg library and project reproducibility
# renv::init()

# load packages
library(BiocManager)
options(repos = BiocManager::repositories()) # Bioconductor repo added to renv
# renv::install("GOSemSim")
# renv::install("AnnotationForge")
# library(GOSemSim)
# library(AnnotationForge)

library(data.table)
library(tinytex)
library(knitr)
library(stringi)
library(tidyverse)

```


#### Read in data tables

``` {r load_and_prep_BUSCO_reports}

# read BFF table
file_BFF <- "busco_reports/full_table_mustela_nigripes.v2.sm.tsv"
full_table_BFF <- as.data.frame(fread(file_BFF, sep = '\t', header = TRUE, fill = TRUE, skip = 4)) %>% 
                  rename( Busco_id = '# Busco id') %>%
                  mutate(Species = 'Mustela nigripes') %>%   # create new col with species name
                  select(Species, everything())    # move new col in front of the other cols
# summary(full_table_BFF)

# read DF table
file_DF <- "busco_reports/full_table_mustela_p_furo_fem_HiC.tsv"
full_table_DF <- as.data.frame(fread(file_DF, sep = '\t', header = TRUE, fill = TRUE, skip = 4)) %>% 
                 rename( Busco_id = '# Busco id') %>%
                 mutate(Species = 'Mustela putorius furo') %>%
                 select(Species, everything())
# summary(full_table_DF)

```


#### Plot number of BUSCOs for each species

``` {r count_genes_and_plot}

# append DF table under BFF table
ferrets <- bind_rows(full_table_BFF,full_table_DF)

# count BUSCOs per species and plot it
ferrets %>%
  group_by(Species) %>%
  count(Status) %>%
  ggplot(aes(x = Status, y = n, fill = factor(Species))) +
    geom_bar(stat = "identity", position = "dodge") +
    scale_fill_grey(name = "Species", 
                    labels = c("Mustela nigripes", "Mustela putorius furo")) +
    labs(title = "BUSCOs found in black-footed ferret and domestic ferret genome assemblies",
         x = "BUSCO gene status", y = "Number of BUSCOs") +
    theme_light() +
    theme(axis.title.x = element_text(size = 12, vjust = -0.90),
          axis.title.y = element_text(size = 12, vjust = 1.90),
          legend.margin = margin(0.5, 0.5, 0.5,0.5)) +
    theme(plot.title = element_text(size = 14,
          margin = margin(0, 0, 6, 0)))

# save barplot
ggsave("Ferrets_all_BUSCOs_barplot.png", width = 16, height = 9, dpi = 300)

```


#### Plot BUSCOs per scaffold 

``` {r count_genes_function}

# declare function for counting genes
count_gene_occurencies_on_scfs <- function(datfr){
  return(datfr %>% 
           group_by(Contig) %>% 
           count(Status) %>% 
           filter(Status != "Missing") %>%  # filter out missing genes, keep others
           arrange(desc(n)))
}

```


#### Plot BUSCOs per scf - BFF

``` {r genes_per_scaffold_BFF}

# call function on BFF gene table and plot counts
full_table_BFF %>%
  count_gene_occurencies_on_scfs() %>%
  ggplot(aes(x = reorder(Contig, n), y = n, fill = Status)) +   # reordering - scaffolds with most genes on top
    geom_bar(stat = "identity", position = "dodge", width = 0.8) +
    scale_fill_grey() +
    coord_flip() +
    labs(title = "BUSCOs found in black-footed ferret scaffolds", 
         x = "Scaffold ID", y = "Number of BUSCOs") +
    theme_light() +
    theme(axis.title.x = element_text(size = 12, vjust = -0.70),
          axis.title.y = element_text(size = 12, vjust = 1),
          legend.margin = margin(0.5, 0.8, 1, 0.5)) +
    theme(plot.title = element_text(size = 14,
          margin = margin(0, 0, 6, 0)))
  
# save barplot
ggsave("BFF_BUSCOs_per_scf_barplot.png", width = 16, height = 9, dpi = 300)
  
```


#### Plot BUSCOs per scf - DF

``` {r genes_per_scaffold_DF}

# call function on DF gene table and plot counts
full_table_DF %>%
  count_gene_occurencies_on_scfs() %>%
  ggplot(aes(x = reorder(Contig, n), y = n, fill = Status)) +  
    geom_bar(stat = "identity", position = "dodge", width = 0.8) +
    scale_fill_grey() +
    coord_flip() +
    labs(title = "BUSCOs found in domestic ferret scaffolds", 
         x = "Scaffold ID", y = "Number of BUSCOs") +
    theme_light() +
    theme(axis.title.x = element_text(size = 12, vjust = -0.70),
          axis.title.y = element_text(size = 12, vjust = 1),
          legend.margin = margin(0.5, 0.8, 1, 0.5)) +
    theme(plot.title = element_text(size = 14,
          margin = margin(0, 0, 6, 0)))

# save barplot
ggsave("DF_BUSCOs_per_scf_barplot.png", width = 16, height = 9, dpi = 300)

```


## Identify shared BUSCOs in both ferret species and unique sets for each species

``` {r join_ferrets_and_orthodb}

# read BUSCO mammalia ortho db with GO terms, protein descriptions,...
file_orthodb <- "mammalia_40674_OrthoDB9_orthogroup_info.txt"
mammalia_orthodb9 <- as.data.frame(read.delim(file_orthodb, sep = '\t', header = TRUE, fill = TRUE)) %>%
                     select(OrthoGroupID, Description, ProtMedianLength, BiologicalProcesses) %>% 
                     rename(Busco_id = OrthoGroupID) # renaming col to enable matching by the same values
# summary(mammalia_orthodb9)

# join orthologs in mammalia db and ferrets based on BUSCO ID, sort by status
ferrets_orthodb_match <- left_join(ferrets, mammalia_orthodb9, by = 'Busco_id') %>%
                         arrange(Status)
  
```


### BUSCOs shared in both ferret species

``` {r get_busco_overlaps_in_both_sp}

# split ferrets in two tables by sp
ortho_match_BFF <- ferrets_orthodb_match %>%
  filter(Species == "Mustela nigripes") 
 
ortho_match_DF <- ferrets_orthodb_match %>%
  filter(Species == "Mustela putorius furo")


# semi join tables by the same BUSCO ID for each status
# semi_join returns rows in 1st table that are matching rows in 2nd table based on given args
shared_buscos_ferrets <- semi_join(ortho_match_BFF, ortho_match_DF, by = c('Busco_id', 'Status')) %>%
                            select(Busco_id, Status, Description, BiologicalProcesses)

```


### BUSCOs unique for each ferret species

``` {r get_unique_buscos_for_each_sp}

# anti_join returns rows in 1st table that have no match in 2nd table based on given args 
# unique buscos in BFF
uniq_buscos_BFF <- anti_join(ortho_match_BFF, ortho_match_DF, by = c('Busco_id', 'Status')) %>%
                      select(Species, Busco_id, Status, Description, BiologicalProcesses)

# unique buscos in DF
uniq_buscos_DF <- anti_join(ortho_match_DF, ortho_match_BFF, by = c('Busco_id', 'Status')) %>%
                     select(Species,Busco_id, Status, Description, BiologicalProcesses)

```


#### Count shared and unique BUSCOs

``` {r count_shared_uniq_buscos}

# declare function 
count_buscos <- function(datfr, type){
  return(datfr %>%
           group_by(Status) %>%
           count(Status)) %>%
           mutate(Type = type) %>%
           select(Type, everything())
}


# count and bind shared and unique buscos in one table
buscos_counts <- bind_rows(
                 count_buscos(shared_buscos_ferrets, "Shared"),
                 count_buscos(uniq_buscos_BFF, "Unique_BFF"),
                 count_buscos(uniq_buscos_DF, "Unique_DF")
)
                  
```


#### Plot shared and unique BUSCOs

``` {r shared_uniq_buscos_plot}

# plot all counts as 1 barplot
buscos_counts %>%
  group_by(Type) %>%
  ggplot(aes(x = Status, y = n, fill = factor(Type))) +
  geom_bar(stat = "Identity", position = "dodge") +
  scale_fill_grey(name = "Species",
                  labels = c("Shared BUSCOs","Unique - M.nigripes", "Unique - M.putorius furo")) +
  labs(title = "BUSCOs shared and unique for each ferret genome assembly",
       x = "BUSCO gene status", y = "Number of BUSCOs") +
  theme_light() +
  theme(axis.title.x = element_text(size = 12, vjust = -0.90),
        axis.title.y = element_text(size = 12, vjust = 1.90),
        legend.margin = margin(0.5, 0.5, 0.5,0.5)) +
    theme(plot.title = element_text(size = 14,
          margin = margin(0, 0, 6, 0)))


# save barplot
ggsave("Ferrets_shared_uniq_BUSCOs_barplot.png", width = 16, height = 9, dpi = 300)  

```


### Duplicated BUSCOs in both species

``` {r shared_busco_duplicates}

# filter out duplicated buscos in both species
BFF_duplicates <- ortho_match_BFF %>% filter(Status == "Duplicated")
DF_duplicates <- ortho_match_DF %>% filter(Status == "Duplicated")

# shared duplicates in both species, regardless of the exact number of duplicates in each sp
shared_dups <- semi_join(BFF_duplicates, DF_duplicates, by = 'Busco_id') %>%
                  select(Busco_id, Status, Description, BiologicalProcesses)

```


#### Unique duplicates in both species 

``` {r unique_duplicates}

# unique dups in BFF
uniq_dups_BFF <- anti_join(BFF_duplicates, DF_duplicates, by = 'Busco_id') %>%
                   select(Species, Busco_id, Status, Description, BiologicalProcesses)

# unique dups in DF
uniq_dups_DF <- anti_join(DF_duplicates, BFF_duplicates, by = 'Busco_id') %>%
                  select(Species,Busco_id, Status, Description, BiologicalProcesses)

```


#### Count shared/unique duplicates 

``` {r count_dups}

# declare function for counting duplicates
count_dups <- function(datfr, type){
  return(datfr %>%
           group_by(Busco_id) %>%
           count(Busco_id)) %>%
           mutate(Type = type) %>%
           select(Type, everything())
}

# count and bind shared and unique dups in one table
buscos_dups <- bind_rows(
                 count_dups(shared_dups, "Shared_duplicates"),
                 count_dups(uniq_dups_BFF, "Unique_BFF_duplicates"),
                 count_dups(uniq_dups_DF, "Unique_DF_duplicates")
)

```

#### Plot shared/unique duplicates 

``` {r plot_dups}

# plot all dups counts
buscos_dups %>%
  group_by(Type) %>%
  summarise(dups_freq_sum = sum(n)) %>%    # sum of all dups for each type(shared/unique)
  ggplot(aes(x = Type, y = dups_freq_sum, fill = factor(Type))) +
  geom_bar(stat = "Identity", position = "dodge") +
  scale_fill_grey(name = "Species",
                  labels = c("Shared duplicates","Unique - M.nigripes", "Unique - M.putorius furo")) +
  labs(title = "Shared and unique duplicated BUSCOs for each ferret genome assembly",
       x = "Type of duplicate", y = "Number of duplicates") +
  theme_light() +
  theme(axis.title.x = element_text(size = 12, vjust = -0.90),
        axis.title.y = element_text(size = 12, vjust = 1.90),
        legend.margin = margin(0.5, 0.5, 0.5,0.5)) +
    theme(plot.title = element_text(size = 14,
          margin = margin(0, 0, 6, 0)))


# save barplot
ggsave("Ferrets_shared_uniq_dups_barplot.png", width = 16, height = 9, dpi = 300)  

```


#### Identifying shared/unique duplicated genes

``` {r identification_of_dups}

# check gene descritions of dups
# get GO.db
# BiocManager::install("GO.db")

# transform ferrets GO ids to terms = transform biol. processes col into list of GO terms
library(GO.db)

# join rows from mamm..db9 to buscos_dups by matches in busco_id col
GO_buscos_dups <- left_join(buscos_dups, mammalia_orthodb9, by = 'Busco_id') %>%
                   dplyr::select(-ProtMedianLength)  # drop col
                  # explicitly called dplyr::select cause it's masked by other pkg's select fun

# call GO.db fun to identify GO IDs




# count NAs per col 
GO_buscos_dups %>% is.na() %>% colSums()

# filter out rows with NAs (missing GO IDs)
no_GO_IDs_dups <- GO_buscos_dups %>% filter_all(any_vars(is.na(.))) 



```




#### Analysis of multiple duplicates

``` {r more_than_2_dups}

# filter out buscos with more than 2 gene copies
BFF_dup_more_than2_cp <- BFF_duplicates %>% group_by(Busco_id) %>% filter(n() > 2)
DF_dup_more_than2_cp <- DF_duplicates %>% group_by(Busco_id) %>% filter(n() > 2)



```

## replace geom_bar(stat=identity) with geom_col 
## analyze duplicates - done
## check shared/unique genes - GO terms

## analyze missing buscos

## check proteinmedianlengths for fragmented and missing buscos
# are these genes much larger and more complex than usual 'completed' gene

## run new busco on cat,dog + ferrets
## get GOs for cat,dog
## compare with ferrets

## bar plot results for each gene set for each sp together, shared/unique separately - package patchwork

## fix plot colors:
#  library(RColorBrewer)
#  display.brewer.all()
#  check theme_bw
#  check theme_minimal
#  theme(legend.position = "top")
#  check fill parameter (does it need to be factored or is it by default)

